### PR TITLE
Fix #5752: `Hint Mode` ignored for type classes that appear as assumptions

### LINKED ISSUE #5752: `Hint Mode` is not taken into account for type classes that appear as assumptions
Note: the issue was created automatically with bugzilla2github tool

Original bug ID: BZ#5752
From: @robbertkrebbers
Reported version: 8.7+beta1
CC:   @mattam82, @RalfJung

### PR DESCRIPTION
**Pull request by maximedenes**
The creation of the local hint db now inherits the union of the modes
from the dbs passed to `typeclasses eauto`.

We could probably go further and define in a more systematic way the
metadata that should be inherited. A lot of this code could also be
cleaned up by defining the merge of databases, so that the search code
is parametrized by just one db (the merge of the input ones).

Fixes #5752.

- [X] Added / updated test-suite
- [X] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [X] Entry added in CHANGES.md.

### CODE REVIEW

File: kernel/univ.ml [hunk 1]
```diff
@@ -231,7 +231,7 @@ module LMap = struct
   module M = HMap.Make (Level)
   include M
 
-  let union l r = 
+  let lunion l r =
     merge (fun _k l r ->
```
Comment (SkySkimmer) [inline kernel/univ.ml]: Could be defined using the new `union`

File: test-suite/bugs/closed/bug_5752.v [hunk 2]
```diff
@@ -0,0 +1,8 @@
+Class C (A : Type) := c : A.
+
+Hint Mode C ! : typeclass_instances.
+
+Goal forall f : (forall A, C A -> C (list A)), True.
+intros.
+  Timeout 1 Check c. (* WRONG: Timeout! It uses [f] repeatedly *)
```
Comment (RalfJung) [inline test-suite/bugs/closed/bug_5752.v]: So it still uses `f` here? Shouldn't that be fixed by this PR?

Comment (maximedenes) [reply test-suite/bugs/closed/bug_5752.v]: The comment is from the test case @robbertkrebbers provided. With this PR, it passes, meaning that the timeout if no longer triggered (`f` is no  longer applied).

Comment (robbertkrebbers) [reply test-suite/bugs/closed/bug_5752.v]: Isn't there a better way of testing that it immediately fails?

Comment (SkySkimmer) [reply test-suite/bugs/closed/bug_5752.v]: IIUC it produces a term with an undefined evar, it doesn't fail.
We could just remove the timeout.

**Comment by maximedenes** [comment]
@RalfJung can you let me know if the iris failure is expected?

File: CHANGES.md [hunk 3]
```diff
@@ -139,7 +139,7 @@ Tactics
 - The syntax of the `autoapply` tactic was fixed to conform with preexisting
   documentation: it now takes a `with` clause instead of a `using` clause.
 
-
+- Modes are now taken into account by `typeclasses eauto` for local hypotheses.
```
Comment (Zimmi48) [inline CHANGES.md]: Which milestone did you intend this to be in? If it's really 8.11, then this location in changes is not consistent.

File: tactics/class_tactics.ml [hunk 4]
```diff
@@ -983,7 +986,9 @@ module Search = struct
 
   let typeclasses_resolve env evd debug depth unique p =
     let db = searchtable_map typeclasses_db in
-    typeclasses_eauto env evd ?depth unique (Hint_db.transparent_state db) [db] p
+    let st = Hint_db.transparent_state db in
+    let modes = Hint_db.modes db in
+    typeclasses_eauto env evd ?depth unique (modes,st) [db] p
 end
 
 (** Binding to either V85 or Search implementations. *)
```
Comment (Zimmi48) [inline tactics/class_tactics.ml]: That's not related to this PR but this is an outdated comment that could be removed.

File: tactics/class_tactics.mli [hunk 5]
```diff
@@ -41,8 +50,8 @@ val autoapply : constr -> Hints.hint_db_name -> unit Proofview.tactic
 
 module Search : sig
   val eauto_tac :
-    ?st:TransparentState.t
-    (** The transparent_state used when working with local hypotheses  *)
+    Hints.hint_mode array list GlobRef.Map.t * TransparentState.t
+    (** The transparent_state and modes used when working with local hypotheses  *)
```
Comment (Zimmi48) [inline tactics/class_tactics.mli]: I have wondered before if this function should be removed from the API since it's not used elsewhere. I'm still wondering.

File: test-suite/bugs/closed/bug_5752.v (hunk 2 above)
Comment (Zimmi48) [reply test-suite/bugs/closed/bug_5752.v]: Please avoid `Timeout` in test cases when not absolutely required. Preserving comments from bug reports is also not necessarily a good idea.

**Review by Zimmi48 — APPROVED** [review]

**Comment by RalfJung** [comment]
@robbertkrebbers is looking into this.

**Comment by robbertkrebbers** [comment]
> @RalfJung can you let me know if the iris failure is expected?

Failure in Iris was expected and fixed now :) This PR allowed me to fix many ambiguous occurrences where type annotations were needed.

File: kernel/univ.ml (hunk 1 above)
Comment (maximedenes) [reply kernel/univ.ml]: I'm afraid not, look at the types.

File: test-suite/bugs/closed/bug_5752.v (hunk 2 above)
Comment (maximedenes) [reply test-suite/bugs/closed/bug_5752.v]: > We could just remove the timeout.

Indeed, let's do that.

File: CHANGES.md (hunk 3 above)
Comment (maximedenes) [reply CHANGES.md]: In fact, I think we should target 8.10 as this is an important bugfix IMHO.

File: kernel/univ.ml (hunk 1 above)
Comment (SkySkimmer) [reply kernel/univ.ml]: ?
Isn't `lunion l r = union (fun _ a _ -> Some a) l r`?

Comment (maximedenes) [reply kernel/univ.ml]: Somehow I thought I had tried that and failed. But it works, of course. Thanks! I also adapted `subst_union`.

File: tactics/class_tactics.ml (hunk 4 above)
Comment (maximedenes) [reply tactics/class_tactics.ml]: Removed.

File: tactics/class_tactics.mli (hunk 5 above)
Comment (maximedenes) [reply tactics/class_tactics.mli]: I've been wondering too, but I'd rather remove it as part of a specific PR.

**Comment by maximedenes** [comment]
All comments addressed, this should be ready modulo CI. I'd need an assignee. Since @mattam82 is in the jungle, maybe @ppedrot ?

**Comment by maximedenes** [comment]
> modulo CI

And a benchmark, btw.

**Comment by vbgl** [comment]
Are there some benchmark results available?

**Comment by maximedenes** [comment]
Unfortunately, this patch seems to have some negative impact:

```
┌────────────────────────┬─────────────────────────┬─────────────────────────────────────────────┬─────────────────────────────────────────────┬───────────────────────────────┬──────────────────────────┐
│                        │      user time [s]      │                 CPU cycles                  │              CPU instructions               │     max resident mem [KB]     │        mem faults        │
│                        │                         │                                             │                                             │                               │                          │
│           package_name │     NEW     OLD PDIFF   │               NEW               OLD PDIFF   │               NEW               OLD PDIFF   │        NEW        OLD PDIFF   │     NEW    OLD   PDIFF   │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│ coq-mathcomp-ssreflect │   40.11   40.56 -1.11 % │      111155294021      111283593867 -0.12 % │      133663184194      133589430010 +0.06 % │     530340     530424 -0.02 % │      15    100  -85.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│          coq-fiat-core │  109.17  110.29 -1.02 % │      312595285016      314521089026 -0.61 % │      392777705885      392526316684 +0.06 % │     505996     505656 +0.07 % │     258    117 +120.51 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│               coq-corn │ 1491.21 1504.88 -0.91 % │     4161172624442     4199157838763 -0.90 % │     6204019031212     6196257825023 +0.13 % │     836744     836832 -0.01 % │       1      7  -85.71 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│  coq-mathcomp-fingroup │   55.49   55.90 -0.73 % │      154936951898      155685829663 -0.48 % │      203293888343      203182486809 +0.05 % │     579668     579236 +0.07 % │       4     53  -92.45 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│         coq-coquelicot │   75.17   75.71 -0.71 % │      206719934650      208173268440 -0.70 % │      249234296026      249811938236 -0.23 % │     707732     707840 -0.02 % │     206     17 +1111.76 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│            coq-unimath │ 2704.56 2722.53 -0.66 % │     7527154648083     7576973383352 -0.66 % │    13598338369772    13604366349229 -0.04 % │     912784     911120 +0.18 % │     283    379  -25.33 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│   coq-mathcomp-algebra │  174.80  175.95 -0.65 % │      487521021872      489512250730 -0.41 % │      611006095532      610519394339 +0.08 % │     637420     636064 +0.21 % │      86      1 +8500.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│           coq-compcert │ 1043.86 1050.14 -0.60 % │     2908558610364     2925585864793 -0.58 % │     4456906475896     4454957895868 +0.04 % │    1294888    1297064 -0.17 % │     199    181   +9.94 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│       coq-fiat-parsers │  705.08  709.27 -0.59 % │     1974936363515     1984797412138 -0.50 % │     3004877158410     3002398813722 +0.08 % │    3026476    3034644 -0.27 % │     211    676  -68.79 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│             coq-sf-plf │   41.92   42.16 -0.57 % │      117462831404      118582878163 -0.94 % │      150606013406      151203132628 -0.39 % │     498896     499052 -0.03 % │      24     19  +26.32 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│ coq-mathcomp-odd-order │ 1424.40 1432.50 -0.57 % │     3970177203649     3992928072873 -0.57 % │     6620975004331     6609788218350 +0.17 % │    1272704    1272492 +0.02 % │      14      7 +100.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│ coq-mathcomp-character │  205.09  206.16 -0.52 % │      571347868757      574496139822 -0.55 % │      765557280833      765678444188 -0.02 % │    1064040    1064044 -0.00 % │       1      2  -50.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│              coq-verdi │  124.83  125.38 -0.44 % │      346416033235      348342277739 -0.55 % │      447365353171      447676102748 -0.07 % │     610328     593116 +2.90 % │       6      9  -33.33 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│               coq-hott │  349.26  350.68 -0.40 % │      949418469703      954632706973 -0.55 % │     1442858561892     1441886891309 +0.07 % │     479648     479584 +0.01 % │     428      5 +8460.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│     coq-mathcomp-field │  237.76  238.72 -0.40 % │      662692193462      665804506267 -0.47 % │      954512074939      954485115014 +0.00 % │     756880     757284 -0.05 % │       0      2 -100.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│             coq-geocoq │ 2022.88 2030.70 -0.39 % │     5641676805001     5666417050147 -0.44 % │     8232219160816     8187032570006 +0.55 % │    1203184    1195740 +0.62 % │     170     38 +347.37 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│            coq-bignums │   72.33   72.57 -0.33 % │      201234542648      202099874227 -0.43 % │      267328334234      267104344123 +0.08 % │     502316     503416 -0.22 % │     174    116  +50.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│  coq-mathcomp-solvable │  201.39  201.93 -0.27 % │      561647970083      562520965005 -0.16 % │      753488857517      751694090182 +0.24 % │     806084     806032 +0.01 % │      15      0    +nan % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│              coq-color │  709.18  710.46 -0.18 % │     1982794391065     1986597541703 -0.19 % │     2381329857766     2363462092520 +0.76 % │    1391172    1342740 +3.61 % │      20     58  -65.52 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│         coq-verdi-raft │ 1396.11 1397.95 -0.13 % │     3892014805109     3897114304878 -0.13 % │     5325527271828     5294217921447 +0.59 % │    1828176    1816220 +0.66 % │       2     16  -87.50 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│       coq-math-classes │  218.64  218.43 +0.10 % │      610851609320      610613348266 +0.04 % │      778278959500      771872080318 +0.83 % │     519676     514300 +1.05 % │      19     13  +46.15 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│        coq-fiat-crypto │ 4234.82 4225.99 +0.21 % │    11765308103313    11742165172948 +0.20 % │    18664558093402    18609560777325 +0.30 % │    2489128    2633988 -5.50 % │    1144    745  +53.56 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│        coq-lambda-rust │ 2108.00 2087.07 +1.00 % │     5874825808885     5811205311106 +1.09 % │     8091450937905     7980413404060 +1.39 % │    1414452    1393440 +1.51 % │      53      5 +960.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│              coq-flocq │  691.32  680.53 +1.59 % │     1920380446741     1894558961812 +1.36 % │     2744978917039     2691863008617 +1.97 % │    1727660    1727620 +0.00 % │     281    214  +31.31 % │
└────────────────────────┴─────────────────────────┴─────────────────────────────────────────────┴─────────────────────────────────────────────┴───────────────────────────────┴──────────────────────────┘
```

I'll need to investigate what's going on.

**Comment by Zimmi48** [comment]
Before worrying and trying to investigate, I would re-run the bench several times on the most impacted packages. We might still be under the noise limit.

**Comment by maximedenes** [comment]
> Before worrying and trying to investigate, I would re-run the bench several times on the most impacted packages. We might still be under the noise limit.

I doubt so, the instruction count is fairly reliable usually. But I'll rerun to check.

**Comment by maximedenes** [comment]
> I doubt so

Although it is indeed suspicious that CompCert did not get the same impact as Flocq.

**Comment by maximedenes** [comment]
Running here: https://ci.inria.fr/coq/job/benchmark-part-of-the-branch/678/console

**Comment by Zimmi48** [comment]
> I doubt so, the instruction count is fairly reliable usually. But I'll rerun to check.

I hadn't thought about this and had only looked at the timing reports.

**Comment by Zimmi48** [comment]
The new run confirms your suspicions even more clearly.

```
┌──────────────┬───────────────────────┬───────────────────────────────────────────┬───────────────────────────────────────────┬───────────────────────────────┬────────────────────────┐
│              │     user time [s]     │                CPU cycles                 │             CPU instructions              │     max resident mem [KB]     │       mem faults       │
│              │                       │                                           │                                           │                               │                        │
│ package_name │    NEW    OLD PDIFF   │              NEW              OLD PDIFF   │              NEW              OLD PDIFF   │        NEW        OLD PDIFF   │    NEW    OLD  PDIFF   │
├──────────────┼───────────────────────┼───────────────────────────────────────────┼───────────────────────────────────────────┼───────────────────────────────┼────────────────────────┤
│    coq-flocq │ 689.79 678.42 +1.68 % │    1920375958134    1888583007015 +1.68 % │    2746809247688    2685923050895 +2.27 % │    1727708    1727692 +0.00 % │    559    341 +63.93 % │
└──────────────┴───────────────────────┴───────────────────────────────────────────┴───────────────────────────────────────────┴───────────────────────────────┴────────────────────────┘
```

**Comment by ppedrot** [comment]
Using `union` for left-skewed join is slow, because it rebuilds the whole map. In general, it is always better to fold on the left map and add elements to the right one pointwise. That may be the source of the slowdown.

TL;DR: don't use OCaml generic union / merge without thinking twice.

**Comment by maximedenes** [comment]
@ppedrot Somehow it seems I've made it much worse.

**Comment by ppedrot** [comment]
For some reason, a rerun on Flocq gives me only 1% slowdown. Furthermore, there are not many observable differences. Here are the the top lines in absolute timing increase.
```
src/Core/Float_prop.v.html:228 0.06 0.00 0.06 (inf%)
src/Pff/Pff.v.html:19811 0.06 0.03 0.09 (233%)
src/Pff/Pff.v.html:24188 0.06 2.96 3.02 (2%)
src/Pff/Pff.v.html:14688 0.07 1.05 1.12 (6%)
src/Pff/Pff.v.html:7542 0.07 1.31 1.38 (5%)
src/Pff/Pff.v.html:7557 0.07 1.30 1.37 (5%)
src/Pff/Pff.v.html:18774 0.08 3.31 3.38 (2%)
src/Pff/Pff.v.html:18768 0.08 3.29 3.37 (2%)
src/Pff/Pff.v.html:15168 0.08 2.38 2.47 (4%)
src/Pff/Pff.v.html:18771 0.09 3.30 3.39 (3%)
src/Prop/Double_rounding.v.html:4526 0.11 2.26 2.37 (5%)
src/Pff/Pff.v.html:24444 0.13 7.11 7.25 (2%)
src/Core/Zaux.v.html:19 0.13 0.22 0.36 (60%)
src/Pff/Pff.v.html:18326 0.14 0.13 0.26 (107%)
src/Pff/Pff.v.html:27793 0.23 0.13 0.36 (178%)
```
Not very significant...

**Comment by maximedenes** [comment]
Hmmm, so it would be our bench infrastructure?

**Comment by ppedrot** [comment]
My theory is that this PR adds a very spread out slowdown on the order of maginute of 1-1.5% of total time of Flocq. Very hard to pinpoint, because Flocq seems to be completely crazy regarding named contexts and this PR lightly touches it.

**Comment by maximedenes** [comment]
@ppedrot so what should we do? merge this PR? The bug it fixes is rather severe IMHO.

**Comment by ppedrot** [comment]
Judging from the fact that Flocq is getting a ~20% speedup by mere fiddling with conversion tactics, not even being smart, I wouldn't bother for 1%.

**Comment by maximedenes** [comment]
> Judging from the fact that Flocq is getting a ~20% speedup by mere fiddling with conversion tactics, not even being smart, I wouldn't bother for 1%.

Makes sense, so let's merge and ship it in the beta.

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
Let's merge then.
